README.md
# Bech32(m) Encoding for Arbitrary Buffers

[![Build status][workflow-image]][workflow-url]
[![Code coverage][coveralls-image]][coveralls-url]
[![Demo][demo-image]][demo-url]

[workflow-image]: https://github.com/slowli/bech32-buffer/workflows/CI/badge.svg?branch=master
[workflow-url]: https://github.com/slowli/bech32-buffer/actions
[coveralls-image]: https://img.shields.io/coveralls/slowli/bech32-buffer.svg
[coveralls-url]: https://coveralls.io/github/slowli/bech32-buffer
[demo-image]: https://img.shields.io/badge/demo-live-blue.svg
[demo-url]: https://slowli.github.io/bech32-buffer/

**Bech32** is a Bitcoin address format specified in [BIP 173][bip-173] and [BIP 350][bip-350].
Among its advantages are: better adaptability to QR codes and in voice conversations,
and improved error detection. This library generalizes Bech32 and its modified version
(Bech32m) to encode any reasonably short byte buffers.

## Usage

### Encoding data

```typescript
declare function encode(
  prefix: string,
  data: Uint8Array,
  encoding: 'bech32' | 'bech32m' = 'bech32'
): string;
```

Encodes binary `data` with the specified human-readable `prefix` into a Bech32(m) string.
The case is preserved: if the prefix is uppercase, then the output will be uppercase
as well; otherwise, the output will be lowercase (including the case when the prefix does
not contain any letters).

#### Arguments

- **prefix:** string  
  Human-readable prefix to hint what kind of data Bech32 encodes. Must contain
  ASCII chars in the range 33-126
- **data:** Uint8Array  
  Binary data to encode
- **encoding:** `bech32` or `bech32m`  
  Specifies whether to use the original Bech32 encoding from [BIP 173][bip-173]
  or the modified encoding from [BIP 350][bip-350]; they differ
  in how a checksum is computed. If omitted, the original encoding is used.

#### Return value

String containing:

1. `prefix`
2. `'1'` separator char
3. `data` encoded with the variant of base32 encoding used by Bech32, and
4. 6-char checksum calculated based on `prefix` and `data`

#### Example

```javascript
const bech32 = require('bech32-buffer');
const data = new Uint8Array(20);
bech32.encode('test', data);
// 'test1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqql6aptf'
```

### Decoding data

```typescript
declare function decode(message: string): {
  prefix: string,
  encoding: 'bech32' | 'bech32m',
  data: Uint8Array
};
```

Extracts human-readable prefix and binary data from the Bech32-encoded string.

#### Arguments

- **data:** string  
  String to decode

#### Return value

An object with the following fields:

- **prefix:** string  
  Human-readable prefix
- **encoding:** `bech32` or `bech32m`  
  Encoding variant inferred from the checksum.
- **data:** Uint8Array  
  Binary data encoded into the input string

Decoding may fail for a variety of reasons (e.g., invalid checksum, or invalid
chars in the input). In this case, `decode()` throws an exception
with a descriptive message.

#### Example

```javascript
const bech32 = require('bech32-buffer');
const data = 'lost1qsyq7yqh9gk0szs5';
bech32.decode(data);
// {
//   prefix: 'lost',
//   encoding: 'bech32',
//   data: Uint8Array([ 4, 8, 15, 16, 23, 42 ])
// }
```

### Bitcoin addresses

```typescript
declare class BitcoinAddress {
  prefix: 'bc' | 'tb';
  scriptVersion: number;
  data: Uint8Array;

  static decode(message: string): BitcoinAddress;
  constructor(prefix: 'bc' | 'tb', scriptVersion: number, data: Uint8Array);
  encode(): string;
  type(): void | 'p2wsh' | 'p2wpkh';
}
```

Provides basic functionality to work with Bech32 encoding of Bitcoin addresses.
Addresses can be `decode`d from strings and `encode`d into strings.
It is also possible to check the `type` of an address. P2WSH and P2WPKH address
types are defined per [BIP 141]. Encoding constraints are defined per [BIP 173][bip-173]
and [BIP 350][bip-350].

#### Example

```javascript
const { BitcoinAddress } = require('bech32-buffer');
const address = BitcoinAddress.decode('BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4');
// address.prefix === 'bc'
// address.scriptVersion === 0
// address.data.length === 20
// address.type() === 'p2wpkh'
```

## Use in Browsers

Use `dist/bech32-buffer.min.js` from the package distribution
or your favorite browserifier. In the first case,
the library will be available as a `bech32` global variable:

```html
<script src="bech32-buffer.min.js"></script>
<!-- later -->
<script>
bech32.encode('test', new Uint8Array(20));
</script>
```

Check out [the web demo](https://slowli.github.io/bech32-buffer/) to see how
**bech32-buffer** works in browser. It is also available in the `examples`
directory of the package.

## Acknowledgements

[BIP 173][bip-173] is authored by Pieter Wuille and Greg Maxwell and is licensed
under the 2-clause BSD license.
[BIP 350][bip-350] is authored by Pieter Wuille and is licensed
under the 2-clause BSD license.

There are at least 2 existing implementations of Bech32 for JavaScript:

- [The reference implementation][ref] by Pieter Wuille
- [Another implementation][bech32] available as the [`bech32` package][bech32-pkg]

Both implementations are Bitcoin-specific, and the reference implementation
is also not in the Npm / yarn package manager.

## License

**bech32-buffer** is available under [Apache-2.0 license](LICENSE).

[bip-173]: https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
[bip-350]: https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki
[ref]: https://github.com/sipa/bech32/tree/master/ref/javascript
[bech32]: https://github.com/bitcoinjs/bech32
[bech32-pkg]: https://www.npmjs.com/package/bech32
[BIP 141]: https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
The project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Internal improvements

- Update demo web page to use newer version of Bootstrap; get rid of jQuery.

## 0.2.1 - 2022-09-27

### Fixed

- Fix encoding with uppercase prefix. Previously, such prefixes led to incorrect
  checksum. With the fix, the prefix case is retained for the encoded string.

## 0.2.0 - 2021-11-21

No substantial changes compared to the 0.2.0-rc.0 release.

## 0.2.0-rc.0 - 2021-11-07

### Added

- Support [Bech32m encoding] for arbitrary data and Bitcoin addresses.
  The new functionality is available on [the demo web page](https://slowli.github.io/bech32-buffer/).

## 0.1.3 - 2021-05-10

### Added

- Document TypeScript typings.

- Update readme to better document main package APIs.

### Internal improvements

- Update dev dependencies.

- Update Bootstrap / JS dependencies for the demo page.

## 0.1.2 - 2018-12-23

### Added

- Add TypeScript typings for the package.

### Internal improvements

- Update dev dependencies.

## 0.1.1 - 2018-12-15

### Added

- Test browser support and add a browser version of the package in the `dist` directory.

- Support Bitcoin addresses via `BitcoinAddress` class.

## 0.1.0 - 2017-07-23

The initial release of `bech32-buffer`.

[Bech32m encoding]: https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki


test/index.js
/* eslint-env node, mocha */

import { expect } from 'chai';
import * as bech32 from '../src';
// Test vectors mentioned in the BIP spec
import vectors from './vectors.json';

const { BitcoinAddress } = bech32;

/**
 * Bitcoin addresses use Bech32 in a peculiar way - script version is
 * not a part of the serialized binary data, but is rather prepended as 5-bit value
 * before the rest of the script. This necessitates some plumbing here.
 *
 * @param {string} prefix
 * @param {number} version
 * @param {Uint8Array} data
 * @returns {string}
 */
function encodeAddress(prefix, version, data) {
  const len = Math.ceil((data.length * 8) / 5);
  const converted = new Uint8Array(len + 1);
  converted[0] = version;
  bech32.to5BitArray(data, converted.subarray(1));
  return bech32.encode5BitArray(prefix, converted);
}

describe('bech32', () => {
  describe('encode', () => {
    vectors.validAddresses.forEach(({ address, script }) => {
      it(`should encode script "${script}" from test vectors`, () => {
        // Extract version from the script
        const prefix = address.substring(0, 2).toLowerCase();
        const scriptData = Buffer.from(script, 'hex');
        const version = scriptData[0] % 32;

        // First 2 bytes of the script are version and the script length,
        // respectively, so they are skipped with `.subarray(2)`.
        expect(encodeAddress(prefix, version, scriptData.subarray(2)))
          .to.equal(address.toLowerCase());
      });
    });

    it('should preserve uppercase prefix', () => {
      const data = new Uint8Array(20);
      const encoded = bech32.encode('T3ST', data);
      expect(encoded).to.match(/^T3ST1[A-Z\d]+$/);

      const { prefix, data: restored } = bech32.decode(encoded);
      expect(prefix).to.equal('t3st');
      expect(restored).to.equalBytes(data);
    });

    it('should fail on mixed-case prefix', () => {
      expect(() => bech32.encode('Test', new Uint8Array(20)))
        .to.throw(/Mixed-case prefix/);
    });

    it('should not encode an overly long message', () => {
      const prefix = 'test';
      const data = new Uint8Array(50);
      // 55 * 8 / 5 = 88 bytes; + 5 bytes for the prefix and separator

      expect(() => bech32.encode(prefix, data)).to.throw(/too long/i);
    });

    it('should not encode an overly long message with long prefix', () => {
      const prefix = 'testtesttesttesttesttesttesttesttesttesttesttesttesttesttest';
      const data = new Uint8Array(20);
      // 20 * 8 / 5 = 32 bytes; + 61 bytes for the prefix and separator

      expect(() => bech32.encode(prefix, data)).to.throw(/too long/i);
    });

    it('should not encode message with invalid chars in prefix', () => {
      const prefix = 'Ю';
      expect(() => bech32.encode(prefix, new Uint8Array(10))).to.throw(/invalid char/i);
    });

    it('should disallow unknown encodings', () => {
      expect(() => bech32.encode('test', new Uint8Array(8), 'bogus'))
        .to.throw(/invalid encoding/i);
    });
  });

  describe('decode', () => {
    vectors.validChecksums.forEach(({ encoded, hrp }) => {
      it(`should decode message with valid checksum "${encoded}"`, () => {
        const decoded = bech32.decode(encoded);
        expect(decoded).to.have.property('prefix', hrp);
        expect(decoded).to.have.property('encoding', 'bech32');
      });
    });

    vectors.validBech32mChecksums.forEach(({ encoded, hrp }) => {
      it(`should decode bech32m message with valid checksum "${encoded}"`, () => {
        const decoded = bech32.decode(encoded);
        expect(decoded).to.have.property('prefix', hrp);
        expect(decoded).to.have.property('encoding', 'bech32m');
      });
    });

    vectors.validAddresses.forEach(({ address, script }) => {
      it(`should decode address "${address}" from test vectors`, () => {
        const decoded = bech32.decodeTo5BitArray(address);
        expect(decoded.prefix).to.satisfy((hrp) => hrp === 'bc' || hrp === 'tb');
        expect(decoded.encoding).to.equal('bech32');

        // Need to remove script version separately, hence `.subarray`
        const decScript = bech32.from5BitArray(decoded.data.subarray(1));

        expect(decScript).to.equalBytes(script.substring(4));
      });
    });

    vectors.validAddresses.forEach(({ bech32mAddress: address, script }) => {
      if (address !== undefined) {
        it(`should decode bech32m address "${address}" from test vectors`, () => {
          const decoded = bech32.decodeTo5BitArray(address);
          expect(decoded.prefix).to.satisfy((hrp) => hrp === 'bc' || hrp === 'tb');
          expect(decoded.encoding).to.equal('bech32m');

          // Need to remove script version separately, hence `.subarray`
          const decScript = bech32.from5BitArray(decoded.data.subarray(1));

          expect(decScript).to.equalBytes(script.substring(4));
        });
      }
    });

    vectors.invalidAddresses.forEach(({ address, reason }) => {
      it(`should detect problem with invalid test address "${address}"`, () => {
        expect(() => {
          const decoded = bech32.decodeTo5BitArray(address);
          bech32.from5BitArray(decoded.data.subarray(1));
        }).to.throw(new RegExp(reason, 'i'));
      });
    });

    vectors.invalidNonAddresses.forEach(({ data, reason }) => {
      it(`should detect problem with "${data}" for reason: ${reason}`, () => {
        expect(() => bech32.decode(data)).to.throw(new RegExp(reason, 'i'));
      });
    });
  });

  describe('BitcoinAddress', () => {
    it('should allow creating custom addresses', () => {
      const data = new Uint8Array(32);
      for (let i = 0; i < data.length; i += 1) {
        data[i] = i;
      }
      const address = new BitcoinAddress('bc', 0, data);
      expect(address.type()).to.equal('p2wsh');

      const encoded = address.encode();
      expect(encoded).to.match(/^bc1/);
      expect(BitcoinAddress.decode(encoded).data).to.equalBytes(data);
    });

    it('should disallow creating addresses with invalid scriptVersion', () => {
      const data = new Uint8Array(32);
      expect(() => new BitcoinAddress('bc', -1, data)).to.throw(RangeError);
      expect(() => new BitcoinAddress('bc', 17, data)).to.throw(RangeError);
    });

    it('should disallow unknown prefixes', () => {
      const data = new Uint8Array(32);
      expect(() => new BitcoinAddress('test', 1, data)).to.throw(/prefix/i);
    });

    it('should disallow overly short scripts', () => {
      const data = new Uint8Array(1);
      expect(() => new BitcoinAddress('tb', 1, data)).to.throw(/script length/i);
    });

    it('should disallow overly long scripts', () => {
      const data = new Uint8Array(41);
      expect(() => new BitcoinAddress('tb', 1, data)).to.throw(/script length/i);
    });

    it('should disallow invalid data lengths for v0 scripts', () => {
      const data = new Uint8Array(24);
      expect(() => new BitcoinAddress('tb', 0, data)).to.throw(/v0 script/i);
    });

    describe('decode', () => {
      vectors.validAddresses.forEach(({ address, bech32mAddress, script }) => {
        const actualAddress = bech32mAddress ?? address;

        it(`should decode address "${actualAddress}" from test vectors`, () => {
          const decoded = BitcoinAddress.decode(actualAddress);
          const scriptData = Buffer.from(script, 'hex');
          const version = scriptData[0] % 32;

          expect(decoded.prefix).to.satisfy((prefix) => prefix === 'tb' || prefix === 'bc');
          expect(decoded.scriptVersion).to.equal(version);
          expect(decoded.data).to.equalBytes(scriptData.subarray(2));
          expect(decoded.encode()).to.equal(actualAddress.toLowerCase());
        });

        if (bech32mAddress != null) {
          it(`should not decode outdated address "${address}" from test vectors`, () => {
            expect(() => BitcoinAddress.decode(address))
              .to.throw(/unexpected encoding/i);
          });
        }
      });

      it('should fail to decode address with invalid prefix', () => {
        expect(() => BitcoinAddress.decode('abcdef1qpzry9x8gf2tvdw0s3jn54khce6mua7lmqqqxw')).to.throw(/prefix/i);
      });

      it('should fail to decode bech32m address for v0 script', () => {
        const address = 'tb1q0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7vq24jc47';
        expect(() => BitcoinAddress.decode(address)).to.throw(/unexpected encoding/i);
      });
    });

    describe('type', () => {
      vectors.validAddresses.forEach(({ address, bech32mAddress, type }) => {
        const actualAddress = bech32mAddress ?? address;

        it(`should guess the type of address "${actualAddress}" (${type})`, () => {
          const decoded = BitcoinAddress.decode(actualAddress);
          expect(decoded.type()).to.equal(type);
        });
      });

      it('should return undefined for unknown v0 scripts', () => {
        const address = BitcoinAddress.decode('BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4');
        address.data = new Uint8Array(36);
        expect(address.type()).to.equal(undefined);
      });
    });
  });
});


test/types.ts
/**
 * This script tests the correspondence of the TypeScript declaration to the actual package.
 * If the TypeScript compiles this file and it can then be successfully executed by Node,
 * we can with some degree of certainty tell that the TS declaration is correct.
 */

import * as chai from 'chai';
import chaiBytes from 'chai-bytes';
import {
  BitcoinAddress,
  decode,
  decodeTo5BitArray,
  encode,
  encode5BitArray,
  FiveBitArray,
  from5BitArray,
  to5BitArray,
} from '..';

// `assert` is easier to type-check than `expect`.
const { assert } = chai.use(chaiBytes);

function testArrayConverter() {
  const buffer = Uint8Array.from([1, 2, 3, 4, 5]);
  const fiveBitsBuffer: FiveBitArray = to5BitArray(buffer);
  assert.lengthOf(fiveBitsBuffer,8);
  const bufferCopy = from5BitArray(fiveBitsBuffer);
  assert.equalBytes(bufferCopy, buffer);
}

function testArrayEncoder() {
  const buffer = Uint8Array.from([1, 2, 3, 4, 5]);
  const fiveBitsBuffer: FiveBitArray = to5BitArray(buffer);
  const encoded: string = encode5BitArray('foo', fiveBitsBuffer);
  assert.isTrue(encoded.startsWith('foo1'));
  const { prefix, data } = decodeTo5BitArray(encoded);
  assert.equal(prefix, 'foo');
  assert.isTrue(data.every((byte, i) => byte === fiveBitsBuffer[i]));
}

function testEncoder() {
  const buffer = Uint8Array.from([0, 1, 2, 3, 4]);
  const encoded: string = encode('test', buffer);
  const { prefix, data: decoded, encoding } = decode(encoded);
  assert.equal(prefix, 'test');
  assert.equal(encoding, 'bech32');
  assert.equalBytes(decoded, buffer);
}

function testModifiedEncoder() {
  const buffer = Uint8Array.from([0, 1, 2, 3, 4]);
  const encoded: string = encode('test', buffer, 'bech32m');
  const { prefix, data: decoded, encoding } = decode(encoded);
  assert.equal(prefix, 'test');
  assert.equal(encoding, 'bech32m');
  assert.equalBytes(decoded, buffer);
}

function testBitcoinAddress() {
  let address = new BitcoinAddress('tb', 0, new Uint8Array(20));
  assert.isTrue((<string> address.encode()).startsWith('tb1'));
  assert.equal(address.data.length, 20);
  assert.equal(address.prefix, 'tb');
  assert.equal(address.scriptVersion, 0);

  address = BitcoinAddress.decode('BC1QW508D6QEJXTDG4Y5R3ZARVARY0C5XW7KV8F3T4');
  assert.equal(address.data.length, 20);
  assert.equal(address.prefix, 'bc');
  assert.equal(address.scriptVersion, 0);
  assert.equal(address.type(), 'p2wpkh');
}

testArrayConverter();
testArrayEncoder();
testEncoder();
testModifiedEncoder();
testBitcoinAddress();


index.d.ts
/**
 * Virtual type for an array in which each element represents 5 bits.
 */
export declare type FiveBitArray = Uint8Array;
// As TypeScript (unlike Flow) always uses structural typing for determining type compatibility,
// there seems to be no way to express this better.

/**
 * Encoding from the Bech32 family used during encoding / decoding.
 */
export declare type Encoding = 'bech32' | 'bech32m';

/**
 * Converts a Uint8Array into a Uint8Array variant, in which each element
 * encodes 5 bits of the original byte array.
 *
 * @param src
 *   Input to convert
 * @param dst
 *   Optional output buffer. If specified, the sequence of 5-bit chunks will be written there;
 *   if not specified, the output buffer will be created from scratch. The length
 *   of `outBuffer` is not checked.
 * @returns
 *   Output buffer consisting of 5-bit chunks
 */
export declare function to5BitArray(src: Uint8Array, dst?: FiveBitArray): FiveBitArray;

/**
 * Converts a sequence of 5-bit chunks into an ordinary Uint8Array.
 *
 * @param src
 *   Input to convert
 * @param dst
 *   Optional output buffer. If specified, the converted bytes will be written there;
 *   if not specified, the output buffer will be created from scratch. The length
 *   of `outBuffer` is not checked.
 * @returns
 *   Output buffer
 */
export declare function from5BitArray(src: FiveBitArray, dst?: Uint8Array): Uint8Array;

/**
 * Encodes binary data into Bech32 encoding.
 *
 * The case is preserved: if the prefix is uppercase, then the output will be uppercase
 * as well; otherwise, the output will be lowercase (including the case when the prefix does
 * not contain any letters).
 *
 * Ordinarily, you may want to use `encode` because it converts
 * binary data to an array of 5-bit integers automatically.
 *
 * @param prefix
 *   Human-readable prefix to place at the beginning of the encoding
 * @param data
 *   Array of 5-bit integers with data to encode
 * @param encoding
 *   Encoding to use; influences the checksum computation. If not specified,
 *   Bech32 encoding will be used.
 * @returns
 *   Bech32(m) encoding of data in the form `<prefix>1<base32 of data><checksum>`
 * @throws If the prefix is mixed-case or contains chars that are not eligible for Bech32 encoding
 */
export declare function encode5BitArray(
  prefix: string,
  data: FiveBitArray,
  encoding?: Encoding
): string;

/**
 * Result of a decoding operation.
 */
export interface DecodeResult<T = Uint8Array> {
  /** Human-readable prefix. */
  prefix: string;
  /** Variation of the Bech32 encoding inferred from the checksum. */
  encoding: Encoding;
  /** Decoded data. */
  data: T;
}

/**
 * Decodes data from Bech32 encoding into an array of 5-bit integers.
 *
 * Ordinarily, you may want to use `decode` because it automatically
 * converts the array of 5-bit integers into an ordinary `Uint8Array`.
 *
 * @param message
 *   Bech32-encoded message
 * @returns
 *   Decoded object with `prefix` and `data` fields, which contain the human-readable
 *   prefix and the array of 5-bit integers respectively.
 */
export declare function decodeTo5BitArray(message: string): DecodeResult<FiveBitArray>;

/**
 * Encodes binary data into Bech32 encoding.
 *
 * The case is preserved: if the prefix is uppercase, then the output will be uppercase
 * as well; otherwise, the output will be lowercase (including the case when the prefix does
 * not contain any letters).
 *
 * @param prefix
 *   Human-readable prefix to place at the beginning of the encoding
 * @param data
 *   Binary data to encode
 * @param encoding
 *   Encoding to use; influences the checksum computation. If not specified,
 *   Bech32 encoding will be used.
 * @returns
 *   Bech32 encoding of data in the form `<prefix>1<base32 of data><checksum>`
 * @throws If the prefix is mixed-case or contains chars that are not eligible for Bech32 encoding
 */
export declare function encode(
  prefix: string,
  data: Uint8Array,
  encoding?: Encoding
): string;

/**
 * Decodes data from Bech32 encoding into an array of 5-bit integers.
 *
 * @param message
 *   Bech32-encoded message
 * @returns
 *   Decoded object with `prefix` and `data` fields, which contain the human-readable
 *   prefix and the decoded binary data respectively.
 */
export declare function decode(message: string): DecodeResult;

/**
 * Bitcoin address.
 */
export declare class BitcoinAddress {
  /**
   * Human-readable prefix. Equals `'bc'` (for mainnet addresses)
   * or `'tb'` (for testnet addresses).
   */
  prefix: 'bc' | 'tb';
  /**
   * Script version. An integer between 0 and 16 (inclusive).
   */
  scriptVersion: number;
  /**
   * Script data. A byte string with length 2 to 40 (inclusive).
   */
  data: Uint8Array;

  /**
   * Decodes a Bitcoin address from a Bech32 string.
   * As per BIP 350, the original encoding is expected for version 0 scripts, while
   * other script versions expect the modified encoding.
   *
   * This method does not check whether the address is well-formed;
   * use `type()` method on returned address to find that out.
   */
  static decode(message: string): BitcoinAddress;

  /**
   * Creates a new address based on provided data.
   *
   * Validation is performed on the fields as specified in their description (e.g.,
   * it is checked that `scriptVersion` is between 0 and 16). Additionally,
   * for `scriptVersion == 0` it is checked that `data` is either 20 or 32 bytes long.
   *
   * @throws {Error} If provided fields do not pass validation.
   */
  constructor(prefix: 'bc' | 'tb', scriptVersion: number, data: Uint8Array);

  /**
   * Encodes this address in Bech32 or Bech32m format, depending on the script version.
   * Version 0 scripts are encoded using original Bech32 encoding as per BIP 173,
   * while versions 1-16 are encoded using the modified encoding as per BIP 350.
   */
  encode(): string;

  /**
   * Guesses the address type based on its internal structure.
   */
  type(): void | 'p2wsh' | 'p2wpkh';
}


src/index.js
/* @flow */

import { toBits, fromBits, createBitArray } from './bit-converter';
import type { BitArray } from './bit-converter';
import {
  expandPrefix,
  CHECKSUM_LENGTH,
  createChecksum,
  verifyChecksum,
  encode as base32Encode,
  decodeWithPrefix,
  detectCase,
} from './encoding';
import type { Encoding } from './encoding';

// Maximum encoded message length
const MAX_ENC_LENGTH = 90;

type FiveBitArray = BitArray<5>;

interface DecodeResult<T = Uint8Array> {
  prefix: string;
  encoding: Encoding;
  data: T;
}

/**
 * Converts a Uint8Array into a Uint8Array variant, in which each element
 * encodes 5 bits of the original byte array.
 *
 * @param {Uint8Array} src
 *   Input to convert
 * @param {?Uint8Array} dst
 *   Optional output buffer. If specified, the sequence of 5-bit chunks will be written there;
 *   if not specified, the output buffer will be created from scratch. The length
 *   of `outBuffer` is not checked.
 * @returns {Uint8Array}
 *   Output buffer consisting of 5-bit chunks
 *
 * @api public
 */
export function to5BitArray(src: Uint8Array, dst?: FiveBitArray): FiveBitArray {
  const len = Math.ceil((src.length * 8) / 5);
  const realDst = dst || createBitArray(len);

  return toBits(src, 5, realDst);
}

export function from5BitArray(src: FiveBitArray, dst?: Uint8Array): Uint8Array {
  const len = Math.floor((src.length * 5) / 8);
  const realDst = dst || new Uint8Array(len);

  return fromBits(src, 5, realDst);
}

/**
 * Encodes binary data into Bech32 encoding.
 *
 * The case is preserved: if the prefix is uppercase, then the output will be uppercase
 * as well; otherwise, the output will be lowercase (including the case when the prefix does
 * not contain any letters).
 *
 * Ordinarily, you may want to use [`encode`](#encode) because it converts
 * binary data to an array of 5-bit integers automatically.
 *
 * @param {string} prefix
 *   Human-readable prefix to place at the beginning of the encoding
 * @param {Uint8Array} data
 *   Array of 5-bit integers with data to encode
 * @param {Encoding} encoding
 *   Encoding to use; influences the checksum computation. If not specified,
 *   Bech32 encoding will be used.
 * @returns {string}
 *   Bech32 encoding of data in the form `<prefix>1<base32 of data><checksum>`
 * @throws If the prefix is mixed-case or contains chars that are not eligible for Bech32 encoding
 *
 * @api public
 */
export function encode5BitArray(
  prefix: string,
  data: FiveBitArray,
  encoding: Encoding = 'bech32',
): string {
  // 1. Allocate buffer for all operations
  const len = 2 * prefix.length + 1 // expanded prefix
    + data.length // five-bit data encoding
    + CHECKSUM_LENGTH; // checksum

  if (len - prefix.length > MAX_ENC_LENGTH) {
    throw new Error(`Message to be produced is too long (max ${MAX_ENC_LENGTH} supported)`);
  }
  const prefixCase = detectCase(prefix, 'prefix') ?? 'lower';

  const buffer = createBitArray<5>(len);

  // 2. Expand the human-readable prefix into the beginning of the buffer
  expandPrefix(prefix.toLowerCase(), buffer.subarray(0, 2 * prefix.length + 1));

  // 3. Copy `data` into the output
  const dataBuffer = buffer.subarray(2 * prefix.length + 1, buffer.length - CHECKSUM_LENGTH);
  dataBuffer.set(data);

  // 4. Create the checksum
  createChecksum(buffer, encoding);

  // 5. Convert into string
  let encoded = base32Encode(buffer.subarray(2 * prefix.length + 1));
  if (prefixCase === 'upper') {
    encoded = encoded.toUpperCase();
  }
  return `${prefix}1${encoded}`;
}

/**
 * Encodes binary data into Bech32 encoding.
 *
 * The case is preserved: if the prefix is uppercase, then the output will be uppercase
 * as well; otherwise, the output will be lowercase (including the case when the prefix does
 * not contain any letters).
 *
 * @param {string} prefix
 *   Human-readable prefix to place at the beginning of the encoding
 * @param {Uint8Array} data
 *   Binary data to encode
 * @param {Encoding} encoding
 *   Encoding to use; influences the checksum computation. If not specified,
 *   Bech32 encoding will be used.
 * @returns {string}
 *   Bech32 encoding of data in the form `<prefix>1<base32 of data><checksum>`
 * @throws If the prefix is mixed-case or contains chars that are not eligible for Bech32 encoding
 *
 * @api public
 */
export function encode(
  prefix: string,
  data: Uint8Array,
  encoding: Encoding = 'bech32',
): string {
  return encode5BitArray(prefix, to5BitArray(data), encoding);
}

/**
 * Decodes data from Bech32 encoding into an array of 5-bit integers.
 *
 * Ordinarily, you may want to use [`decode`](#decode) because it automatically
 * converts the array of 5-bit integers into an ordinary `Uint8Array`.
 *
 * @param {string} message
 *   Bech32-encoded message
 * @returns {DecodeResult<FiveBitArray>}
 *   Decoded object with `prefix` and `data` fields, which contain the human-readable
 *   prefix and the array of 5-bit integers respectively.
 *
 * @api public
 */
export function decodeTo5BitArray(message: string): DecodeResult<FiveBitArray> {
  // Check preconditions

  // 1. Message length
  if (message.length > MAX_ENC_LENGTH) {
    throw new TypeError(`Message too long; max ${MAX_ENC_LENGTH} expected`);
  }

  // 2. Mixed case
  detectCase(message); // we don't care about the result, only about checks.
  const lowerCaseMsg = message.toLowerCase();

  // 4. Existence of the separator char
  const sepIdx = lowerCaseMsg.lastIndexOf('1');
  if (sepIdx < 0) {
    throw new Error('No separator char ("1") found');
  }

  // 5. Placing of the separator char in the message
  if (sepIdx > message.length - CHECKSUM_LENGTH - 1) {
    throw new Error(`Data part of the message too short (at least ${CHECKSUM_LENGTH} chars expected)`);
  }

  const prefix = lowerCaseMsg.substring(0, sepIdx);

  // Checked within `decodeWithPrefix`:
  // 6. Invalid chars in the data part of the message
  const bitArray = decodeWithPrefix(prefix, lowerCaseMsg.substring(sepIdx + 1));

  // 7. Checksum
  const encoding = verifyChecksum(bitArray);
  if (encoding === undefined) {
    throw new Error('Invalid checksum');
  }

  return {
    prefix,
    encoding,
    // Strip off the prefix from the front and the checksum from the end
    data: bitArray.subarray(2 * prefix.length + 1, bitArray.length - CHECKSUM_LENGTH),
  };
}

/**
 * Decodes data from Bech32 encoding into an array of 5-bit integers.
 *
 * @param {string} message
 *   Bech32-encoded message
 * @returns {DecodeResult}
 *   Decoded object with `prefix` and `data` fields, which contain the human-readable
 *   prefix and the decoded binary data respectively.
 *
 * @api public
 */
export function decode(message: string): DecodeResult<> {
  const { prefix, encoding, data: bitArray } = decodeTo5BitArray(message);
  return { prefix, encoding, data: from5BitArray(bitArray) };
}

/**
 * Bitcoin address.
 */
export class BitcoinAddress {
  /**
   * Human-readable prefix. Equal to `'bc'` (for mainnet addresses)
   * or `'tb'` (for testnet addresses).
   */
  prefix: 'bc' | 'tb';

  /**
   * Script version. An integer between 0 and 16 (inclusive).
   */
  scriptVersion: number;

  /**
   * Script data. A byte string with length 2 to 40 (inclusive).
   */
  data: Uint8Array;

  /**
   * Decodes a Bitcoin address from a Bech32(m) string.
   * As per BIP 350, the original encoding is expected for version 0 scripts, while
   * other script versions expect the modified encoding.
   *
   * This method does not check whether the address is well-formed;
   * use `type()` method on returned address to find that out.
   *
   * @param {string} message
   * @returns {BitcoinAddress}
   */
  static decode(message: string): this {
    const { prefix, data, encoding } = decodeTo5BitArray(message);

    // Extra check to satisfy Flow.
    if (prefix !== 'bc' && prefix !== 'tb') {
      throw new Error('Invalid human-readable prefix, "bc" or "tb" expected');
    }

    const scriptVersion = data[0];
    if (scriptVersion === 0 && encoding !== 'bech32') {
      throw Error(`Unexpected encoding ${encoding} used for version 0 script`);
    }
    if (scriptVersion > 0 && encoding !== 'bech32m') {
      throw Error(`Unexpected encoding ${encoding} used for version ${scriptVersion} script`);
    }
    return new this(prefix, scriptVersion, from5BitArray(data.subarray(1)));
  }

  constructor(prefix: 'bc' | 'tb', scriptVersion: number, data: Uint8Array) {
    if (prefix !== 'bc' && prefix !== 'tb') {
      throw new Error('Invalid human-readable prefix, "bc" or "tb" expected');
    }
    if ((scriptVersion < 0) || (scriptVersion > 16)) {
      throw new RangeError('Invalid scriptVersion, value in range [0, 16] expected');
    }
    if (data.length < 2 || data.length > 40) {
      throw new RangeError('Invalid script length: expected 2 to 40 bytes');
    }
    if ((scriptVersion === 0) && (data.length !== 20 && data.length !== 32)) {
      throw new Error('Invalid v0 script length: expected 20 or 32 bytes');
    }

    this.prefix = prefix;
    this.scriptVersion = scriptVersion;
    this.data = data;
  }

  /**
   * Guesses the address type based on its internal structure.
   *
   * @returns {void | 'p2wpkh' | 'p2wsh'}
   */
  type(): void | 'p2wpkh' | 'p2wsh' {
    if (this.scriptVersion !== 0) {
      return undefined;
    }

    switch (this.data.length) {
      case 20: return 'p2wpkh';
      case 32: return 'p2wsh';

      // should be unreachable, but it's JS, so you never know
      default: return undefined;
    }
  }

  /**
   * Encodes this address in Bech32 or Bech32m format, depending on the script version.
   * Version 0 scripts are encoded using original Bech32 encoding as per BIP 173,
   * while versions 1-16 are encoded using the modified encoding as per BIP 350.
   *
   * @returns {string}
   *   Bech32(m)-encoded address
   */
  encode(): string {
    // Bitcoin addresses use Bech32 in a peculiar way - script version is
    // not a part of the serialized binary data, but is rather prepended as 5-bit value
    // before the rest of the script. This necessitates some plumbing here.
    const len = Math.ceil((this.data.length * 8) / 5);
    const converted: FiveBitArray = createBitArray(len + 1);
    converted[0] = this.scriptVersion;
    to5BitArray(this.data, converted.subarray(1));

    const encoding = (this.scriptVersion === 0) ? 'bech32' : 'bech32m';
    return encode5BitArray(this.prefix, converted, encoding);
  }
}
